### ISSUE TITLE
Make "completed initial" default behavior when points are entered for non-lead jammer

### ISSUE THREAD

**Issue by AdamSmasherDerby** [rollerderby/scoreboard#428]
Based on actual .json files from current users, it is a real problem to remember to enter the "initial pass" for non-lead jamemrs.  I am seeing half a dozen or more errors related to this in pretty much every game I'm getting sent.

Can we improve this with training?  Probably.

Should we FIX this in software?  Definitely.  

Points on an initial pass should still be possible to enter, but other than overtime, it should take effort to DO so, not effort to NOT do so.  They are a screamingly near-impossible edge case, rather than something that happens every game, and our default behaviors should match what actually occurs in gameplay.

**Comment by brian-brazil** [comment]
>  it is a real problem to remember to enter the "initial pass" for non-lead jamemrs

Even if we change this, the data entered will still be wrong in the case that the initial trip was completed but no points scored on the first trip - which is a very common case.

Papering over one place operators fail to do complete data entry doesn't fix the actual problem, it only superficially makes things seem better and hides a signal of data quality.

> They are a screamingly near-impossible edge case

The scoreboard targets rulesets beyond WFTDA.

**Comment by Capt-Brunch** [comment]
I honestly do not have information for additional rulesets, but I can confirm that the "score + NI" or "score + more score" behavior for the initial pass seems problematic. If a game is created wherein WFTDA rules are selected, I would expect that the scoreboard behaves as if it were using the WFTDA ruleset and thus remove NI when points are scored, and also stop allowing up to two trips worth of points on the first trip.

**Comment by brian-brazil** [comment]
The WFTDA ruleset permits points in the intial trip. We have no rulesets that affect stats tracking, and I don't think we should start adding them.

Version 4 of the scoreboard requires that operators correctly track trips if they want the scoreboard to display correct information, and if they want correct stats to be recorded.

**Comment by AdamSmasherDerby** [comment]
The idea that we should PUNISH operators for making mistakes, rather than trying to account for them, is not my ideal of user interface design. 

The fact that there may be extreme edge cases that allow for points on the initial trip (and I remain unconvinced that there are) does not mean that behavior should be the default.

Even assuming a perfect operator who never makes mistakes, you want to make the usual case the simple one to enter just for ease of use.

**Comment by brian-brazil** [comment]
> The idea that we should PUNISH operators for making mistakes

Uhm, where has anyone said anything about punishment on behalf of the scoreboard? There is no functionality anywhere in the scoreboard designed with punishing users in mind, that's a very strange idea to me.

This discussion seems to be going quite a bit off track, and beyond technical issues. If you feel that the burden placed on the operator to press one additional button per typical jam compared to v3 is too high a cost for full SK paperwork, you can always use an older version and continue to rely on paper.

**Comment by Capt-Brunch** [comment]
I can agree with Adam's statement that this feels like "punishing" behavior in the sense that, for what I fully expect to be more than 99.9% of jams, you're requiring the scoreboard operator to complete an additional click/keypress that's only necessary in very rare scenarios, and which results in inaccuracies if not done. This is a UX issue far more than it is a technical one.

There are two paths here - the first, which we're currently on, requires an extra click/keypress in virtually all scenarios in exchange for saving that click in extreme edge cases. The other path, which Adam and I are arguing should be the default, prevents the need for that click in the overwhelming majority of scenarios and only requires it in the rare instance in which a specific and rare scenario were to occur.

It surprises me that you're advocating for additional onus on scoreboard operator to account for edge cases, when you've argued in the past that edge cases should be the instances to require extra interactions as opposed to typical cases.

**Comment by frank-weinberg** [comment]
Even if this feature request were implemented, there are still reasons that the SBO should mark initial trip completed:

1. The majority of jams out in the wild end x-0 with the non-lead Jammer on a scoring trip. For these jams this feature request does not change a thing. So the SBO will still have to press a button at some point in order to get correct stats.
2. We are displaying on the main scoreboard when a Jammer is (marked to be) on their initial trip.
3. We are also collecting duration of (initial) trips with the aim of generating stats for benches and announcers from them (think Rainy bench stats). For these stats to be useful the end of the initial trip needs to be marked reasonably close to when it happened.

So in order to properly operate CRG 4 an input has to be made when the initial trip is completed and there is no way around that.

I have trained a double digit number of people on CRG 4 and none of them have taken more than 2 games before they handed in correct paperwork. Also none of them complained that it was too much of a hassle to fix it when they occasionally forgot to mark the end of an initial trip. So I believe the issues Smasher is currently seeing will significantly reduce as people get better trained and get more routine with CRG 4. (Let's remember that it has only been released for 2 months of off-season so far.)

So this feature request would be incentivising people to operate the scoreboard in a way that causes the collected data to be incorrect while taking away the biggest indicator to the operator that they are doing so.

**Comment by AdamSmasherDerby** [comment]
There's two cases here:

The non-lead jammer scores points.
The non-lead jammer does not score points.

You seem to be saying that to give people an incentive to handle the second case correctly, we are going to ALWAYS handle the first case wrong by default.

That is bad UI design.  

We don't turn off autocorrect for some words on our phones in order to incentivize people to learn to spell. We should have the software do as much of the work as possible, and if people are getting NI vs. 0 point trips wrong, THAT'S what you fix with education.  Not the thing we can fix ourselves in software.

**Comment by frank-weinberg** [comment]
No. I am saying that even with the proposed change both cases will still be handled wrong. The wrongness will just be more subtle in the first case, making it harder for people to become aware of their mistakes and fix them before they become a habit.

**Comment by official-sounding** [comment]
The scoreboard already advances to a scoring trip when Lead is marked.  That is technically incorrect behavior per the rules (Lead is awarded before the initial trip is completed, a Lead Jammer ending up with NI is rare, but unlike points on an initial trip, it can actually happen outside an officiating error). But, it is a major convenience factor that prefers the common case over corner cases or "making people aware of their mistakes".  I think logic for the case for similar behavior when adding points to a non-lead Jammer is pretty much the same logic that leads to the Lead case.

Based on the discussion here and in the developers group, I am assuming you guys won't be convinced to implement this.  If I make a PR for it, is it going to be an acrimonious fight, or will it be relatively smooth sailing?

**Comment by brian-brazil** [comment]
The difference I see with Lead is that it's unlikely that they won't eventually get out by the end of the jam, and it saves you a button press and in that typical case is more likely to produce correct stats. We're basically saying that Lead and the initial trip usually happen at the same time, and optimise for that case.

> I think logic for the case for similar behavior when adding points to a non-lead Jammer is pretty much the same logic that leads to the Lead case.

That's not the case here though. Completing the initial trip and getting a score are going to be separated by at least 10 seconds (a very fast lap&trip) and failing to record the initial trip by a non-lead jammer will produce incorrect stats in the very common x-0 score case.

Any changes proposed here need to address how we ensure that NI gets correctly tracked in the x-0 case. As it stands always pressing it immediately when the initial is complete seems the easiest to do/teach, and the most reliable to me.

You also want NI tracked correctly in real time for dot system reasons - namely helping you to remember whether the jammer is on initial when they return from a penalty. Even more so if there's no paper.

My experiences line up with Speedy's in relation to the effectiveness of training.

> If I make a PR for it, is it going to be an acrimonious fight, or will it be relatively smooth sailing?

Both of us are in agreement here, so it seems very unlikely such a PR would end up being accepted.

**Comment by official-sounding** [comment]
> As it stands always pressing it immediately when the initial is complete seems the easiest to do/teach, and the most reliable to me.

I don't think anyone is arguing that this is the ideal scenario - train officials to hit the button immediately after ending a trip.  This is about handling cases when users forget to do the ideal thing. Doing the actively wrong thing when they add points instead of the thing they almost certainly meant to do so they can be "aware of their mistakes" is user hostile to me. It's not a perfect safety net, but I don't want perfect to be the enemy of the good.

Additionally, from a training/habits perspective I would think all zero-point jams being NI is still a sufficient "red flag" to allow trainers to identify when operators are not being trained correctly.

It's fine if y'all do not want to implement the code on this, but since you're already willing to maintain "optimizations" to help operators when a case is "unlikely", I don't see why you wouldn't accept a properly written PR with unit tests, etc.  to help with a different optimization that helps operators.

**Comment by brian-brazil** [comment]
It's not a case of perfect being an enemy of good (that'd be if for example we removed the current lead logic), it'd be a case of adding complexity to produce one wrong set of stats instead of a different wrong set of stats. Let's focus instead on what we need to do to get the right stats for all jams.

There's nothing stopping this being post processed by Smasher's tool or similar after a game (hopefully with a warning about implausible stats), but I don't think such a thing belongs in the scoreboard itself.

> Additionally, from a training/habits perspective I would think all zero-point jams being NI is still a sufficient "red flag" to allow trainers to identify when operators are not being trained correctly.

Definitely, it's one of the things I personally look for on paper when training someone in addition to no NI on SP (possible but extremely rare) or no trailing 0s. It's not as in your face as the initial trip score notation though, so is more likely to be missed. Not that that's my primary argument, but it's an additional reason that this is the better approach.

> since you're already willing to maintain "optimizations" to help operators when a case is "unlikely"

I think you may have misread what I wrote. The lead case deals with the likely case, I'd estimate that only for 2-3% of jams would what we do with lead be incorrect in terms of ultimate stats.

**Comment by AdamSmasherDerby** [comment]
If you want to talk about percentages, what we are doing now when points are entered for the not lead jammer without the trip being manually ended is going to be wrong in 99.9999999.... percent of cases.

If we can safely assume user intent, (and we can, as the alternative is as close to impossible as makes no difference) we should.

**Comment by frank-weinberg** [comment]
Let's face reality: A large part of our user base does not read documentation and also does not have ready access to high level training.

I firmly believe that in this environment it is worse UI design to give the user the impression that everything is handled correctly when it actually isn't than it is to not fix the part of the error that we could fix.

So if you want to implement this, you have to come up with a different way of making the user aware that there are errors in the data that couldn't be fixed. There also needs to be an exception for overtime and a way to turn off the behaviour for game types where points during initial are more common.

And in my experience an operator who has gotten used to the new interface will miss marking initial complete before entering points less than once per game on average. So all this effort will save less than 2 clicks per game. I don't think that's worth it but if someone else is willing to put in the work, I can see myself accepting a PR if the above points are adequately addressed (and it meets the usual requirements on code quality etc.).

**Comment by brian-brazil** [comment]
> A large part of our user base does not read documentation

I work on the presumption that no one reads it. It's too dense, technical, and referency given our userbase.

> also does not have ready access to high level training.

They have access to the videos I made, which clearly cover this topic in the basic videos.

> I don't think that's worth it but if someone else is willing to put in the work, I can see myself accepting a PR if the above points are adequately addressed 

I'm not sure we should be adding so much complexity purely to save two clicks per game. Complexity isn't free, and I think getting it down to two clicks (you don't want to know how you had to fix it before I made it only two clicks) is where we should stop in terms of tradeoffs.

I also have difficulty seeing how you'd address these points. Interrupting the user to inform them that they made a mistake would not be the best thing to be doing for an inexperienced operator. We also need to expose such a warning to the HNSO, not just the operator.

I think this better belongs as an external post-processing step if it's really desired.
